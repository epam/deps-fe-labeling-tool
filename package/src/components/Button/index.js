import { Button, ButtonType } from './Button'
import { IconButton } from './IconButton'
import { TextButton } from './TextButton.styles'

Button.Icon = IconButton

export {
  Button,
  ButtonType,
  TextButton
}
