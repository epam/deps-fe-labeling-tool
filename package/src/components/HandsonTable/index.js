
export * from './models'
export * from './HandsonTable'
