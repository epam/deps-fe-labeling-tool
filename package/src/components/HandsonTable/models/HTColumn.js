
import PropTypes from 'prop-types'

class HTColumn {
  constructor (data) {
    this.data = data
  }
}

const htColumnShape = PropTypes.shape({
  data: PropTypes.string.isRequired
})

export {
  HTColumn,
  htColumnShape
}
