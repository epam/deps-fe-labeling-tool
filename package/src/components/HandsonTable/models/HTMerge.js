import PropTypes from 'prop-types'

class HTMerge {
  constructor (row, col, rowspan, colspan) {
    this.row = row
    this.col = col
    this.rowspan = rowspan
    this.colspan = colspan
  }
}

const htMergeShape = PropTypes.shape({
  row: PropTypes.number.isRequired,
  col: PropTypes.number.isRequired,
  rowspan: PropTypes.number.isRequired,
  colspan: PropTypes.number.isRequired
})

export {
  HTMerge,
  htMergeShape
}
