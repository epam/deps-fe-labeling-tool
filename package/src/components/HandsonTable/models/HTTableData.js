
import PropTypes from 'prop-types'

const rowOfPrimitivesShape = PropTypes.arrayOf(
  PropTypes.oneOfType([
    PropTypes.number.isRequired,
    PropTypes.string.isRequired
  ])
)

const rowsOfPrimitivesShape = PropTypes.arrayOf(
  rowOfPrimitivesShape
)

export { rowsOfPrimitivesShape }
