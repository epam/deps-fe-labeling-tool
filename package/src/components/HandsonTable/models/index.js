
export * from './HTTableData'
export * from './HTMerge'
export * from './HTColumn'
