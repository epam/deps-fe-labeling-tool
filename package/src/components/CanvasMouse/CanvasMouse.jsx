import { useEffect } from 'react'
import PropTypes from 'prop-types'
import { canvasShape } from '@/models/Canvas'

const CanvasMouse = (props) => {
  useEffect(() => {
    props.onMouseDown && props.canvas.on('mouse:down', props.onMouseDown)
    props.onMouseMove && props.canvas.on('mouse:move', props.onMouseMove)
    props.onMouseUp && props.canvas.on('mouse:up', props.onMouseUp)
    return () => {
      props.onMouseDown && props.canvas.off('mouse:down', props.onMouseDown)
      props.onMouseMove && props.canvas.off('mouse:move', props.onMouseMove)
      props.onMouseUp && props.canvas.off('mouse:up', props.onMouseUp)
    }
  },
  [
    props.canvas,
    props.onMouseDown,
    props.onMouseMove,
    props.onMouseUp
  ]
  )

  return null
}

CanvasMouse.propTypes = {
  canvas: canvasShape.isRequired,
  onMouseDown: PropTypes.func,
  onMouseMove: PropTypes.func,
  onMouseUp: PropTypes.func
}

export {
  CanvasMouse
}
