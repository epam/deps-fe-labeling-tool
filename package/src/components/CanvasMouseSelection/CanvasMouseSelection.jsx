import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { CanvasMouse } from '@/components/CanvasMouse'
import { withCanvas } from '@/components/CanvasProvider'
import { canvasShape } from '@/models/Canvas'
import { Rectangle } from '@/models/Rectangle'
import { COLORS } from '@/theme/theme.default'

const MIN_REQUIRED_THRESHOLD = 0.001

class CanvasMouseSelection extends PureComponent {
  getSelectionRectangle = (opts) => {
    const cursorB = this.props.canvas.getPointer(opts.e)
    if (!this.cursorA || !cursorB) {
      return
    }

    const w = Math.abs(this.cursorA.x - cursorB.x)
    const h = Math.abs(this.cursorA.y - cursorB.y)
    const x = Math.min(this.cursorA.x, cursorB.x)
    const y = Math.min(this.cursorA.y, cursorB.y)

    if (w === 0 || h === 0) {
      const width = Math.max(w, MIN_REQUIRED_THRESHOLD)
      const height = Math.max(h, MIN_REQUIRED_THRESHOLD)
      return new Rectangle(x, y, width, height)
    }

    return new Rectangle(x, y, w, h)
  }

  onMouseDown = (opts) => {
    if (this.props.canvas.findTarget(opts.e) || opts.e.altKey) {
      return
    }

    this.cursorA = this.props.canvas.getPointer(opts.e)
  }

  onMouseMove = (opts) => {
    const selection = this.getSelectionRectangle(opts)
    selection && this.props.onSelectionUpdate && this.props.onSelectionUpdate(
      selection,
      opts
    )
  }

  onMouseUp = (opts) => {
    const selection = this.getSelectionRectangle(opts)
    selection && this.props.onSelectionEnd(
      selection,
      opts
    )

    this.cursorA = undefined
  }

  componentDidMount = () => {
    this.props.canvas.selectionColor = COLORS.SELECTION
    this.props.canvas.selectionBorderColor = COLORS.SELECTION_BORDER_COLOR
  }

  componentWillUnmount = () => {
    this.props.canvas.selectionColor = COLORS.TRANSPARENT
    this.props.canvas.selectionBorderColor = COLORS.TRANSPARENT
  }

  render = () => (
    <CanvasMouse
      canvas={this.props.canvas}
      onMouseDown={this.onMouseDown}
      onMouseMove={this.props.onSelectionUpdate && this.onMouseMove}
      onMouseUp={this.onMouseUp}
    />
  )
}

CanvasMouseSelection.propTypes = {
  canvas: canvasShape.isRequired,
  onSelectionEnd: PropTypes.func.isRequired,
  onSelectionUpdate: PropTypes.func
}

const WithCanvas = withCanvas(CanvasMouseSelection)

export {
  WithCanvas as CanvasMouseSelection
}
