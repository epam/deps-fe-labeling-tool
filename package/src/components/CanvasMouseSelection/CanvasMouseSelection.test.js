import { mockCanvasProvider } from '@/mocks/mockCanvasProvider'
import { MockCanvas, mockFabric } from '@/mocks/mockFabric'
import React from 'react'
import { shallow } from 'enzyme'
import { CanvasMouse } from '@/components/CanvasMouse'
import { CanvasMouseSelection as ProviderMock } from '@/components/CanvasMouseSelection'
import { Rectangle } from '@/models/Rectangle'
import { COLORS } from '@/theme/theme.default'

jest.mock('fabric', () => mockFabric)

jest.mock('@/components/CanvasProvider', () => mockCanvasProvider)

const mockCanvas = new MockCanvas()
mockCanvas.getPointer = jest.fn(() => ({
  x: 10,
  y: 10
}))

const MIN_REQUIRED_THRESHOLD = 0.001
const mockRectangle = new Rectangle(10, 10, 10, 10)
const mockOpts = { e: { altKey: null } }
const mockCursor = {
  x: 20,
  y: 20
}

const CanvasMouseSelection = ProviderMock.WrappedComponent

describe('Component: CanvasMouseSelection', () => {
  let defaultProps
  let wrapper

  beforeEach(() => {
    defaultProps = {
      canvas: mockCanvas,
      onSelectionEnd: jest.fn(),
      onSelectionUpdate: jest.fn()
    }

    wrapper = shallow(<CanvasMouseSelection {...defaultProps} />)
  })

  it('should render correct layout', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should set selection color on component mount', () => {
    expect(defaultProps.canvas.selectionColor).toBe(COLORS.SELECTION)
    expect(defaultProps.canvas.selectionBorderColor).toBe(COLORS.SELECTION_BORDER_COLOR)
  })

  it('should set selection color to transparent on component unmount', () => {
    wrapper.instance().componentWillUnmount()
    expect(defaultProps.canvas.selectionColor).toBe(COLORS.TRANSPARENT)
    expect(defaultProps.canvas.selectionBorderColor).toBe(COLORS.TRANSPARENT)
  })

  it('should call canvas getPointer on onMouseDown event', () => {
    wrapper.find(CanvasMouse).props().onMouseDown(mockOpts)
    expect(defaultProps.canvas.getPointer).nthCalledWith(1, mockOpts.e)
  })

  it('should not call canvas getPointer on onMouseDown event when given altKey', () => {
    const opts = { e: { altKey: 'altKey' } }
    wrapper.find(CanvasMouse).props().onMouseDown(opts)
    expect(defaultProps.canvas.getPointer).not.toHaveBeenCalled()
  })

  it('should call defaultProps.onSelecionEnd on onMouseUp event', () => {
    wrapper.instance().cursorA = mockCursor
    wrapper.find(CanvasMouse).props().onMouseUp(mockOpts)
    expect(defaultProps.onSelectionEnd).nthCalledWith(1, mockRectangle, mockOpts)
  })

  it('should call defaultProps.onSelectionUpdate on onMouseMove event', () => {
    wrapper.instance().cursorA = mockCursor
    wrapper.find(CanvasMouse).props().onMouseMove(mockOpts)
    expect(defaultProps.onSelectionUpdate).nthCalledWith(1, mockRectangle, mockOpts)
  })

  it('CanvasMouseSelection.getSelectionRectangle should return rectangle', () => {
    wrapper.instance().cursorA = mockCursor
    expect(wrapper.instance().getSelectionRectangle(mockOpts)).toEqual(mockRectangle)
  })

  it('CanvasMouseSelection.getSelectionRectangle should return rectangle with minimum size if coords are equal', () => {
    wrapper.instance().cursorA = mockCanvas.getPointer()
    expect(wrapper.instance().getSelectionRectangle(mockOpts)).toEqual(new Rectangle(10, 10, MIN_REQUIRED_THRESHOLD, MIN_REQUIRED_THRESHOLD))
  })

  it('CanvasMouseSelection.getSelectionRectangle should return undefined if missing cursor coords', () => {
    expect(wrapper.instance().getSelectionRectangle(mockOpts)).toEqual(undefined)
  })
})
