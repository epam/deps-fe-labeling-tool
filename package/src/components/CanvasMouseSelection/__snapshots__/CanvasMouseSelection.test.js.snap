// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CanvasMouseSelection should render correct layout 1`] = `
<CanvasMouse
  canvas={
    MockCanvas {
      "add": [MockFunction],
      "backgroundColor": "#ffffff",
      "backgroundImage": Object {},
      "discardActiveObject": [MockFunction],
      "findTarget": [MockFunction],
      "fire": [MockFunction],
      "fxRemove": [MockFunction],
      "getActiveObject": [MockFunction],
      "getActiveObjects": [MockFunction],
      "getCenter": [MockFunction],
      "getHeight": [MockFunction],
      "getObjects": [MockFunction],
      "getPointer": [MockFunction],
      "getWidth": [MockFunction],
      "getZoom": [MockFunction],
      "height": 768,
      "listeners": Object {},
      "off": [MockFunction],
      "on": [MockFunction],
      "ref": undefined,
      "remove": [MockFunction],
      "renderAll": [MockFunction],
      "requestRenderAll": [MockFunction],
      "selection": true,
      "selectionBorderColor": "rgba(255, 255, 255, 0.3)",
      "selectionColor": "rgba(133, 165, 255, 0.2)",
      "sendToBack": [MockFunction],
      "setActiveObject": [MockFunction],
      "setBackgroundColor": [MockFunction],
      "setBackgroundImage": [MockFunction],
      "setCursor": [MockFunction],
      "setHeight": [MockFunction],
      "setViewportTransform": [MockFunction],
      "setWidth": [MockFunction],
      "setZoom": [MockFunction],
      "viewportTransform": Array [
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "width": 1366,
      "zoom": 1,
      "zoomToPoint": [MockFunction],
    }
  }
  onMouseDown={[Function]}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
/>
`;
